### PR TITLE
[Issue-1509] reimplemented `is_widgets_enabled` function

### LINKED ISSUE #1509: [BUG] `deepchecks.utils.ipython.is_widgets_enabled` does not work properly
**Short summary**

regarding `deepchecks.utils.ipython.is_widgets_enabled` function

That is not the right way to verify whether widgets are enabled or not:
- there is always a possibility that a user had started the jupyter server with not default config path
- there are two extension types:
     + classical notebook extensions (managed by 'jupyter nbextension');
     + jupyterlab extensions (managed by 'jupyter labextension');

### PR DESCRIPTION
**Pull request by yromanyshyn**
resolves #1509 

NOTE: still testing, and writing tests

### CODE REVIEW

File: deepchecks/utils/ipython.py [hunk 1]
```diff
@@ -95,28 +99,8 @@ def is_headless() -> bool:
...
-            # disabled and enabled, regard it as disabled
-            output = subprocess.getoutput('jupyter nbextension list').split('\n')
-            disabled_regex = re.compile(r'\s*(jupyter-js-widgets/extension).*(disabled).*')
-            enabled_regex = re.compile(r'\s*(jupyter-js-widgets/extension).*(enabled).*')
-            found_disabled = any((disabled_regex.match(s) for s in output))
-            found_enabled = any((enabled_regex.match(s) for s in output))
-            return not found_disabled and found_enabled
-        except Exception:  # pylint: disable=broad-except
-            return False
+    warnings.warn('', category=DeprecationWarning)
```
[the remark this answers is not in the document]
Comment (yromanyshyn) [reply deepchecks/utils/ipython.py]: renamed back

File: deepchecks/utils/ipython.py [hunk 2]
```diff
@@ -304,121 +267,540 @@ def __exit__(self, *args, **kwargs):
...
+    warning_kwargs.update(**kwargs)
+    return _EXTENSION_WARNING.format(**warning_kwargs)
+
+
+def is_widgets_enabled() -> bool:
+    """Find out whether ipywidgets use is possible within jupyter interactive REPL."""
+    is_jupyterlab_enabled = is_jupyter_server_extension_enabled('jupyterlab')
+    is_nbclassic_enabled = is_jupyter_server_extension_enabled('nbclassic')
+
+    if is_jupyterlab_enabled and is_nbclassic_enabled:
```
Comment (matanper) [inline deepchecks/utils/ipython.py:704]: same here about duplicate code

**Comment by matanper** [comment]
I worry that users that have both classic and lab installed will have too many warnings. how often do we think warnings will appear?

File: deepchecks/utils/ipython.py [hunk 3]
```diff
@@ -304,121 +267,540 @@ def __exit__(self, *args, **kwargs):
...
+
+        return is_plotly_enabled and is_widgets_enabled
+
+    elif is_nbclassic_enabled:
+        is_extension_enabled = is_nbclassic_extension_enabled('jupyter-js-widgets')
+
+        if is_extension_enabled is False:
+            warnings.warn(_nbclassic_widgets_warning(additional=''))
+
+        return is_extension_enabled
```
[the remark this answers is not in the document]
Comment (yromanyshyn) [reply deepchecks/utils/ipython.py]: fixed

**Comment by yromanyshyn** [comment]
@matanper 

> I worry that users that have both classic and lab installed will have too many warnings. how often do we think warnings will appear?

I added filters for that, warnings will appear only once when `is_widgets_enabled` or `is_interactive_output_use_possible` is called for the first time

File: deepchecks/utils/ipython.py [hunk 4]
```diff
@@ -304,121 +267,511 @@ def __exit__(self, *args, **kwargs):
...
+        # TODO:
+        # I did not find any good instruction on internet
+        # probably better to write our own instruction for this
+        install_instructions='+ https://pypi.org/project/jupyter-js-widgets-nbextension/\n'
+    )
+    warning_kwargs.update(**kwargs)
+    return _EXTENSION_WARNING.format(**warning_kwargs)
+
+
+def is_widgets_enabled() -> bool:
```
Comment (JKL98ISR) [inline deepchecks/utils/ipython.py:699]: this function is doubled now

Comment (JKL98ISR) [inline deepchecks/utils/ipython.py:699]: seems like a lot of duplicated code

Comment (JKL98ISR) [reply deepchecks/utils/ipython.py:699]: is_widgets_enabled and is_interactive_output_use_possible

Comment (yromanyshyn) [reply deepchecks/utils/ipython.py:699]: they are doing different things
1. verifies whether the use of interactive outputs is possible (whether nbclassic/jupyterlab widgets and jupyterlab plotly extensions are enabled)
3. verifies whether the use of ipywidgets is possible (whether nbclassic/jupyterlab widgets extension are enabled)

Comment (matanper) [reply deepchecks/utils/ipython.py:699]: what's the difference between interactive output and ipywidgets? is there another kind of interactive output?

Comment (JKL98ISR) [reply deepchecks/utils/ipython.py:699]: still many of this code. is duplicated

Comment (yromanyshyn) [reply deepchecks/utils/ipython.py:699]: > what's the difference between interactive output and ipywidgets? is there another kind of interactive output?

under interactive output I mean, situation when we use plotly and ipywidgets to display check/suite results

Comment (yromanyshyn) [reply deepchecks/utils/ipython.py:699]: >still many of this code is duplicated

sorry I do not get you
could you be more specific, please?

Comment (matanper) [reply deepchecks/utils/ipython.py:699]: so `is_interactive_output_use_possible` checks plotly + widgets and `is_widgets_enabled` checks only widgets?

Comment (yromanyshyn) [reply deepchecks/utils/ipython.py:699]: > so is_interactive_output_use_possible checks plotly + widgets and is_widgets_enabled checks only widgets?

yes,
we use `ipywidgets` for progress bars and they definitely do not require `jupyterlab-plotly` extension to work

File: deepchecks/core/suite.py [hunk 5]
```diff
@@ -141,7 +141,7 @@ def _ipython_display_(
                 ),
                 raw=True
             )
-        elif is_widgets_enabled() and as_widget and not is_kaggle_env():
+        elif not is_kaggle_env() and is_interactive_output_use_possible() and as_widget:
```
Comment (matanper) [inline deepchecks/core/suite.py:144]: reminder why we do `not is_kaggle_env()`?

Comment (JKL98ISR) [reply deepchecks/core/suite.py:144]: plotly widgets aren't working in kaggle even tho it seems like kaggle has widgets, maybe with this update we don't need it?
@yromanyshyn can you check if it's still needed?

File: deepchecks/utils/ipython.py [hunk 6]
```diff
@@ -304,121 +267,511 @@ def __exit__(self, *args, **kwargs):
...
+
+def extract_jupyter_server_token(url: str) -> str:
+    """Extract token string from jupyter server url query params string."""
+    query = parse_qs(url)
+    token = (query.get('token') or [])
+    return (token[0] if len(token) > 0 else '')
+
+
+def is_jupyterlab_extension_enabled(name: str) -> bool:
+    """Find out whether provided nbclassic extension is enabled."""
```
Comment (matanper) [inline deepchecks/utils/ipython.py:373]: why nbclassic?

File: deepchecks/utils/ipython.py [hunk 7]
```diff
@@ -304,121 +267,511 @@ def __exit__(self, *args, **kwargs):
...
+    server_url = get_jupyter_server_url()
+    extensions = None
+
+    if server_url is not None:
+        extensions = request_jupyterlab_extensions(server_url)
+
+    if extensions is None:
+        extensions = get_jupyterlab_extensions()
```
Comment (matanper) [inline deepchecks/utils/ipython.py:381]: is this a more reliable way? why not just always use this and remove the server request?

File: deepchecks/utils/ipython.py (hunk 6 above)
Comment (yromanyshyn) [reply deepchecks/utils/ipython.py:373]: oh, sorry made a mistake
will fix

File: deepchecks/utils/ipython.py (hunk 7 above)
Comment (yromanyshyn) [reply deepchecks/utils/ipython.py:381]: cli command will always take a look at the default config path but
what if the user started a server with some non-default config path?

File: deepchecks/core/suite.py (hunk 5 above)
Comment (yromanyshyn) [reply deepchecks/core/suite.py:144]: I do not remember exactly/
whether they needed `jupyterlab-plotly` extension and it was not installed
or whether there was an incompatible version of it installed

yup I will check

Comment (yromanyshyn) [reply deepchecks/core/suite.py:144]: image
image

`go.FigureWidget` does not work on kaggle
this happens only for ploty>5.5.0

File: deepchecks/utils/ipython.py (hunk 6 above)
Comment (yromanyshyn) [reply deepchecks/utils/ipython.py]: fixed

File: deepchecks/utils/ipython.py (hunk 7 above)
Comment (matanper) [reply deepchecks/utils/ipython.py:381]: Please add explanation about that in the code

Comment (yromanyshyn) [reply deepchecks/utils/ipython.py:381]: added

**Pull request closed without merging**